proxmoxtf/resource/vm/vm.go

```diff
@@ -3029,6 +3107,49 @@ func vmGetHostPCIDeviceObjects(d *schema.ResourceData) vms.CustomPCIDevices {
 	return pciDeviceObjects
 }
 
+func vmGetNumaDeviceObjects(d *schema.ResourceData) vms.CustomNUMADevices {
+	numaNode := d.Get(mkNUMA).([]interface{})
+	numaNodeObjects := make(vms.CustomNUMADevices, len(numaNode))
+
+	for i, numaNodeEntry := range numaNode {
+		block := numaNodeEntry.(map[string]interface{})
+
+		deviceName := block[mkNUMADevice].(string)
+		ids := block[mkNUMACPUIDs].(string)
+		hostNodes, _ := block[mkNUMAHostNodeNames].(string)
+		memory, _ := block[mkNUMAMemory].(int)
+		policy, _ := block[mkNUMAPolicy].(string)
+
+		device := vms.CustomNUMADevice{
+			Memory: &memory,
+			Policy: &policy,
+		}
+
+		if ids != "" {
+			dIDs := strings.Split(ids, ";")
+			device.CPUIDs = dIDs
+		}
+
+		if hostNodes != "" {
+			dHostNodes := strings.Split(hostNodes, ";")
+			device.HostNodeNames = &dHostNodes
+		}
+
+		if strings.HasPrefix(deviceName, "numa") {
+			deviceID, err := strconv.Atoi(deviceName[4:])
+			if err == nil {
+				numaNodeObjects[deviceID] = device
+
+				continue
+			}
+		}
+
+		numaNodeObjects[i] = device
```

Comment:
@bpg  I did't find how you use `device` in `hostpci` block, but I've added some non restricted index code.

Reply:
I must admit, the [code](https://github.com/bpg/terraform-provider-proxmox/blob/31b6812ce281da214261786f633ed7e93c197fb0/proxmoxtf/resource/vm/vm.go#L3691C1-L3695C3) in hostpci devices does not really work as required.

The idea is to avoid VM update if order of the blocks in plan changes, so when each block is identified by `device` value, the provider can find a match in the remote state and reconcile them.
But the local state also need to ignore order changes.

Anyway, I made necessary changes, and apply seems to be working as expected, i.e. manually reordering blocks after apply is not detected as a resource update.

The new code will also help to fix #185 and similar issues.